### PR TITLE
Allow filtering within the Operator Namespace

### LINKED ISSUE #138: Filtering does not work on the operator namespace
Filtering introduced in #132 does not filter in the operator namespace. The motivation for this is that we want to extract all resources from this namespace to extract as much information as possible that pertains to the operator. 

However this becomes a problem when users deploy workloads into the operator namespace (something that is discouraged but not explicitly forbidden, we need to call this out more explicitly in  documentation). 

To remedy this problem we should rework the way we treat the operator namespace: 
* we should respect any filters defined by the user in the operator namespace
* we should still extract operator related resources from the operator namespace but be more targeted e.g. try to identify the operator Pod and extract its logs
* we need to distinguish between different deployment modes of the operator (there is some initial code around this already in the tool) to cover OLM installations as well as manual YAML  manifest based installations and Helm based installs.

### PR DESCRIPTION
**Pull request by naemono**
closes #138 

Allows filtering of resources within the Operator namespace only when resources are deployed within the Operator namespace.

Tested with
- [x] Helm deployed eck-operator with and without resources deployed within the same namespace
- [x] Yaml manifest deployed eck-operator with and without resources deployed within the same namespace

### CODE REVIEW

File: internal/diag.go [hunk 1]
```diff
@@ -142,13 +152,44 @@ func Run(params Params) error {
...
+			"common.k8s.elastic.co/type=maps",              // 1.6.0
+		}
+
+		if slices.Contains(params.OperatorNamespaces, ns) {
+			// If the current resource namespace is present within the set of
+			// operator namespaces then find the label in use by operator and
+			// add this to the set of filters.
+			//
+			// 1) if using yaml manifests it will always be "control-plane=elastic-operator"
+			// 2) if using Helm, find label with key: helm.sh/chart, and value containing "eck-operator-*"
```
Comment (thbkrkr) [inline internal/diag.go:179]: Nit: I would move this info into the `getOperatorLabel` godoc.

File: internal/diag.go [hunk 2]
```diff
@@ -232,6 +264,29 @@ LOOP:
...
 }
 
+func getOperatorLabel(kc *Kubectl, ns string) (labels.Set, error) {
+	saList, err := kc.Clientset.CoreV1().ServiceAccounts(ns).List(context.Background(), v1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("while retrieving list of serviceaccounts in ns '%s': %w", ns, err)
+	}
+	if saList == nil {
+		return nil, fmt.Errorf("nil sa while retrieving list of serviceaccounts in ns '%s'", ns)
+	}
```
Comment (thbkrkr) [inline internal/diag.go:274]: This is not really necessary as we can iterate over a nil slice and then we will return `fmt.Errorf("unable to find serviceaccount corresponding to eck operator")` which has a slightly more informative message IMO.

```suggestion
```

**Review by thbkrkr — COMMENTED** [review]
I've a slightly different approach to propose.

Until this PR, we had two ways of running a diagnostic in a namespace depending if it is an operator namespace or a resources namespace. Now we say, if a namespace is both, we can skip the "operator diag" and just run the "resources diag" because it also collects diagnostic information about the operator. At this point, I think we can keep only the "resources diag".

This is what you implemented in pseudo-code (`runOperatorDiag` and `runResourcesDiag` doesn't exist). 

```go
for _, ns := range params.OperatorNamespaces {
        // if we detect that the operator namespace has also been declared as a resources namespace,
        // we don't run the "operator diag"
	if slices.Contains(params.ResourcesNamespaces, ns) {
		continue
	}
	runOperatorDiag()
}

for _, ns := range params.ResourcesNamespaces {
	runResourcesDiag()
}

func runResourcesDiag(..., selector) {
	...
        // if this an operator and resources namespace we add the operator labelSelector
	if slices.Contains(params.OperatorNamespaces, ns) {
		namespaceFilters = params.Filters.WithSelector(selector)
		logsLabels = append(logsLabels, selector.String())
	}
	...
}
```

Proposal:

```go
for _, ns := range params.OperatorNamespaces {
	// remove the operator namespace from the resources namespaces if present
	params.ResourcesNamespaces = remove(params.ResourcesNamespaces, ns)

	// run the resources diag with the operator label selector
	operatorlabelSelector := getOperatorLabel(kubectl, ns)
	runDiag(..., operatorlabelSelector)
}

for _, ns := range params.ResourcesNamespaces {
	runDiag(..., nil)
}

func runDiag(params Params, zipFile *archive.ZipFile, stopCh chan struct{}, kubectl *Kubectl, maxOperatorVersion *version.Version, ns string, selector labels.Selector) error {
	...
	// add the optional labelSelector
	if labelSelector != nil {
		namespaceFilters = params.Filters.WithSelector(labelSelector)
		logsLabels = append(logsLabels, labelSelector.String())
	}
	...
}
```

To implement this, we need to create `runDiag` and call it for the two types of namespaces.

**Comment by pebrc** [comment]
I wonder if we can go even one step further and drop the whole special treatment of the operator namespaces with the exception of the logs for the operator. 

```golang
diff --git a/internal/diag.go b/internal/diag.go
index b38a313..c8caf42 100644
--- a/internal/diag.go
+++ b/internal/diag.go
@@ -9,6 +9,7 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"os"
 	"os/signal"
 	"path/filepath"
@@ -112,46 +113,32 @@ func Run(params Params) error {
 	})
 
 	operatorVersions := make([]*version.Version, 0, len(params.OperatorNamespaces))
-
 	for _, ns := range params.OperatorNamespaces {
 		operatorVersions = append(operatorVersions, detectECKVersion(clientSet, ns, params.ECKVersion))
 
-		// If this operator namespace is also included in the resources namespace,
-		// just continue as it will be handled within the resource namespaces loop.
-		if slices.Contains(params.ResourcesNamespaces, ns) {
-			continue
+		label, err := getOperatorLabel(kubectl, ns)
+		if err != nil {
+			return err
 		}
-
-		logger.Printf("Extracting Kubernetes diagnostics from %s\n", ns)
-		zipFile.Add(getResources(kubectl.GetByLabel, ns, filters.Empty, []string{
-			"statefulsets",
-			"pods",
-			"services",
-			"configmaps",
-			"events",
-			"networkpolicies",
-			"controllerrevisions",
-			"serviceaccount",
-		}))
-
-		zipFile.Add(map[string]func(io.Writer) error{
-			archive.Path(ns, "secrets.json"): func(writer io.Writer) error {
-				return kubectl.GetMeta("secrets", ns, writer)
-			},
-		})
+		if label == nil {
+			logger.Printf("Could not find label corresponding to ECK Operator: potentially not including operator data in diagnostics")
+		}
+		namespaceFilters := params.Filters.WithSelector(label.AsSelector())
+		logsLabels := label.AsSelector().String()
 
 		// Filters is intentionally empty here, as label filtering doesn't apply to
 		// pods in the operator namespace.
-		if err := kubectl.Logs(ns, "", filters.Empty, zipFile.Create); err != nil {
-			zipFile.AddError(err)
-		}
+		getLogs(kubectl, zipFile, ns, namespaceFilters, logsLabels)
 	}
 
 	maxOperatorVersion := max(operatorVersions)
 	logVersion(maxOperatorVersion)
 
+	allNamespaces := sets.New[string](params.ResourcesNamespaces...)
+	allNamespaces.Insert(params.OperatorNamespaces...)
+
 LOOP:
-	for _, ns := range params.ResourcesNamespaces {
+	for _, ns := range allNamespaces {
 		namespaceFilters := params.Filters
 		select {
 		case <-stopCh:

```

We would just always add the operator namespaces to the resource namespace with the only special treatment for the operator namespaces being that we add extract the logs for the operator separately.

**Comment by thbkrkr** [comment]
> We would just always add the operator namespaces to the resource namespace with the only special treatment for the operator namespaces being that we add extract the logs for the operator separately.

@pebrc in your version, I still see the special treatment for a resource namespace that is also an operator namespace:

https://github.com/elastic/eck-diagnostics/blob/0f7e961afb32ff065e8a6544e06837a2ddaadbaf/internal/diag.go#L173-L189

We need to add the operator label selector not only for logs but also for k8s resources (e.g.: StatefulSet) in case there is a elastic resource filter (`-f elasticsearch=wxyz`), otherwise we don't get the operator StatefulSet but that of wxyz es cluster .

**Comment by naemono** [comment]
@thbkrkr @pebrc I have taken both of your comments into account, and made appropriate changes, and ran some manual testing and all looks good. I also made a change to ensure multiple operators in different namespaces are taken into account, as I hadn't handled that possibility previously.

File: internal/diag.go [hunk 3]
```diff
@@ -106,49 +112,56 @@ func Run(params Params) error {
...
+			"common.k8s.elastic.co/type=agent",             // 1.4.0
+			"common.k8s.elastic.co/type=maps",              // 1.6.0
+		}
+
+		selectors := make([]labels.Selector, len(operatorLabels))
+		for i, label := range operatorLabels {
+			selectors[i] = label.AsSelector()
+			logsLabels = append(logsLabels, label.AsSelector().String())
+		}
+		namespaceFilters := params.Filters.WithSelectors(selectors)
```
Comment (thbkrkr) [inline internal/diag.go:161]: I think we can now simplify a bit more by moving this block outside of the loop and add the operator label selector to `params.Filters` and `logsLabel` on the fly when iterating through operator namespaces.

**Review by pebrc — APPROVED** [review]
LGTM

File: internal/diag.go (hunk 3 above)
Comment (thbkrkr) [reply internal/diag.go]: We could even avoid constructing and iterating over `operatorLabels` but I think it's a little clearer that way.

**Pull request merged**